Add support for runtime.sendMessage from Web Extension content scripts.

xeenon opened the pull request:
#### 75bd7acd2bff1ab8cf96511d44cca4a8f690e658
<pre>
Add support for runtime.sendMessage from Web Extension content scripts.
<a href="https://webkit.org/b/261909">https://webkit.org/b/261909</a>
rdar://problem/115864149

Reviewed by Brent Fulgham.

* Adds support for runtime.sendMessage and main world firing of onMessage events.
* Tracks event listeners by type and what world they are registered in, so we can get
  the appropriate processes to dispatch the events.
* Added better support for the sender object and tests for the values it includes.

* Source/WebKit/DerivedSources-input.xcfilelist: Added WebExtensionContentWorldType.serialization.in.
* Source/WebKit/DerivedSources.make: Added WebExtensionContentWorldType.serialization.in.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared): Added WebExtensionContentWorldType.
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
(WebKit::WebExtensionContext::addListener): Added contentWorldType.
(WebKit::WebExtensionContext::removeListener): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm: Added.
(WebKit::WebExtensionContext::runtimeSendMessage): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsSendMessage):Added more sender parameters.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::processes const): Added contentWorldType.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::pageListensForEvent const): Added contentWorldType.
(WebKit::WebExtensionContext::processes const): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::sendToProcessesForEvent): Pass Main for contentWorldType. This
is only used for main world events currently and avoided touching a bunch of call sites.
(WebKit::WebExtensionContext::sendToContentScriptProcessesForEvent): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added WebExtensionContentWorldType.*.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::addListener): Added contentWorldType.
(WebKit::WebExtensionAPIEvent::removeListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::sendMessage): Added.
(WebKit::toWebAPI): Moved here since it is only used here.
(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent): .
(WebKit::WebExtensionContextProxy::dispatchRuntimeMainWorldMessageEvent): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::sendMessage): Added more sender parameters.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::addListener): Added contentWorldType.
(WebKit::WebExtensionAPIWebNavigationEvent::removeListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
(WebKit::WebExtensionAPIWindowsEvent::addListener): Added contentWorldType.
(WebKit::WebExtensionAPIWindowsEvent::removeListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h:
(WebKit::WebExtensionAPIObject::contentWorldType const): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/268317@main">https://commits.webkit.org/268317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6465859c411e797b551b738ec1178817eb731e59

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19326 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19746 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21217 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18084 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19869 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19545 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16804 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22086 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/19503 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16780 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17592 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17842 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21897 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18359 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15545 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17492 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21852 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2362 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

davidjohnson91 commented on Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h:
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
Should this copyright date be 2023?

xeenon replied:
I'll get it right one of these years. Maybe next year…

brentfulgham reviewed (approved):
LGTM (with the date correction). r=me

xeenon committed:
Add support for runtime.sendMessage from Web Extension content scripts.
https://webkit.org/b/261909
rdar://problem/115864149

Reviewed by Brent Fulgham.

* Adds support for runtime.sendMessage and main world firing of onMessage events.
* Tracks event listeners by type and what world they are registered in, so we can get
  the appropriate processes to dispatch the events.
* Added better support for the sender object and tests for the values it includes.

* Source/WebKit/DerivedSources-input.xcfilelist: Added WebExtensionContentWorldType.serialization.in.
* Source/WebKit/DerivedSources.make: Added WebExtensionContentWorldType.serialization.in.
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared): Added WebExtensionContentWorldType.
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionContentWorldType.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionMessageSenderParameters.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIAlarmsCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIEventCocoa.mm:
(WebKit::WebExtensionContext::addListener): Added contentWorldType.
(WebKit::WebExtensionContext::removeListener): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIRuntimeCocoa.mm: Added.
(WebKit::WebExtensionContext::runtimeSendMessage): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITabsCocoa.mm:
(WebKit::WebExtensionContext::tabsSendMessage):Added more sender parameters.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::processes const): Added contentWorldType.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::pageListensForEvent const): Added contentWorldType.
(WebKit::WebExtensionContext::processes const): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::sendToProcessesForEvent): Pass Main for contentWorldType. This
is only used for main world events currently and avoided touching a bunch of call sites.
(WebKit::WebExtensionContext::sendToContentScriptProcessesForEvent): Added.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added WebExtensionContentWorldType.*.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIEventCocoa.mm:
(WebKit::WebExtensionAPIEvent::addListener): Added contentWorldType.
(WebKit::WebExtensionAPIEvent::removeListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIRuntimeCocoa.mm:
(WebKit::WebExtensionAPIRuntime::sendMessage): Added.
(WebKit::toWebAPI): Moved here since it is only used here.
(WebKit::WebExtensionContextProxy::dispatchRuntimeMessageEvent): .
(WebKit::WebExtensionContextProxy::dispatchRuntimeMainWorldMessageEvent): Added.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPITabsCocoa.mm:
(WebKit::WebExtensionAPITabs::sendMessage): Added more sender parameters.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWebNavigationEventCocoa.mm:
(WebKit::WebExtensionAPIWebNavigationEvent::addListener): Added contentWorldType.
(WebKit::WebExtensionAPIWebNavigationEvent::removeListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsEventCocoa.mm:
(WebKit::WebExtensionAPIWindowsEvent::addListener): Added contentWorldType.
(WebKit::WebExtensionAPIWindowsEvent::removeListener): Ditto.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIObject.h:
(WebKit::WebExtensionAPIObject::contentWorldType const): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIRuntime.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIRuntime.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIRuntime.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/268317@main
